pkg/kubelet/network/cni/cni.go

```diff
@@ -0,0 +1,232 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cni
+
+import (
+	"fmt"
+	"github.com/appc/cni/libcni"
+	cniTypes "github.com/appc/cni/pkg/types"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/kubelet/network"
+	kubeletTypes "k8s.io/kubernetes/pkg/kubelet/types"
+	"net"
+	"sort"
+)
+
+const (
+	CNIPluginName        = "cni"
+	DefaultPluginName    = "kubernetes-bridge"
+	DefaultNetDir        = "/etc/cni/net.d"
+	DefaultCNIDir        = "/opt/cni/bin"
+	VendorCNIDirTemplate = "/opt/%s/bin"
+)
+
+type cniNetworkPlugin struct {
+	cniNetworkMap map[string]*cniNetwork
+	host          network.Host
+}
+
+type cniNetwork struct {
+	name          string
+	NetworkConfig *libcni.NetworkConfig
+	CNIConfig     *libcni.CNIConfig
+}
+
+func ProbeNetworkPlugins() []network.NetworkPlugin {
+	configList := make([]network.NetworkPlugin, 0)
+	return append(configList, &cniNetworkPlugin{cniNetworkMap: getAllCNINetworks()})
+}
+
+func getAllCNINetworks() map[string]*cniNetwork {
+	networkMap := make(map[string]*cniNetwork)
+	files, err := libcni.ConfFiles(DefaultNetDir)
+	switch {
+	case err != nil:
+		return networkMap
+	case files == nil || len(files) == 0:
+		return networkMap
+	}
+
+	sort.Strings(files)
+	for _, confFile := range files {
+		conf, err := libcni.ConfFromFile(confFile)
+		if err != nil {
+			glog.Warningf("Error loading CNI config file %s: %v", confFile, err)
+			continue
+		}
+		// Search for vendor-specific plugins as well as default plugins in the CNI codebase.
+		vendorCNIDir := fmt.Sprintf(VendorCNIDirTemplate, conf.Network.Name)
+		cninet := &libcni.CNIConfig{
+			Path: []string{DefaultCNIDir, vendorCNIDir},
+		}
+		networkMap[conf.Network.Name] = &cniNetwork{name: conf.Network.Name, NetworkConfig: conf, CNIConfig: cninet}
+	}
+	return networkMap
+}
+
+func (plugin *cniNetworkPlugin) selectNetwork(pod *api.Pod) (*cniNetwork, error) {
+	// loading networks all over again maybe inefficient, but required if networks can be created on the fly
+	plugin.cniNetworkMap = getAllCNINetworks()
+
+	if len(plugin.cniNetworkMap) == 0 {
+		return nil, fmt.Errorf("No available CNI network available in %s", DefaultNetDir)
+	}
+
+	// check if the namespace of the pod is a network name itself
+	network, ok := plugin.cniNetworkMap[pod.Namespace]
+	if ok {
+		return network, nil
+	}
+
+	// label called "network"?
+	network, ok = plugin.cniNetworkMap[pod.Labels["network"]]
+	if ok {
+		return network, nil
+	}
+
+	// annotation called "network"?
+	network, ok = plugin.cniNetworkMap[pod.Annotations["network"]]
+	if ok {
+		return network, nil
+	}
+
+	// else return the first one
+	for _, network := range plugin.cniNetworkMap {
```

Comment:
maps iterate in semi-random order.  If you have 2 networks this will not choose the same one each time.  You'll need a different structure.  But to answer a later question, I think joining them all is correct for now.